syntax = "proto3";
package orderbook;

import "google/protobuf/timestamp.proto";

service OrderBookService {
    rpc place_order         (OrderRequest)           returns (OrderResponse);
    rpc cancel_order        (CancelOrderRequest)     returns (OrderResponse);
    rpc get_order_book      (GetOrderBookRequest)    returns (OrderBookResponse);
    rpc get_order_status    (GetOrderStatusRequest)  returns (OrderResponse);
    rpc get_trade_history   (GetTradeHistoryRequest) returns (TradeHistoryResponse);
    rpc place_orders        (OrderBatchRequest)      returns (OrderBatchResponse);
}

message OrderRequest {
    uint64 id = 1;
    string price = 2;
    string quantity = 3;
    Side side = 4;
    OrderType order_type = 5;
}

message OrderResponse {
    uint64 id = 1;
    string price = 2;
    string quantity = 3;
    string remaining_quantity = 4;
    Side side = 5;
    OrderType order_type = 6;
    OrderStatus status = 7;
    google.protobuf.Timestamp timestamp = 8;
}

message CancelOrderRequest {
    uint64 order_id = 1;
}

message GetOrderBookRequest {
    uint32 depth = 1;
}

message OrderBookLevel {
    string price = 1;
    string quantity = 2;
}

message OrderBookResponse {
    repeated OrderBookLevel bids = 1;
    repeated OrderBookLevel asks = 2;
}

message GetOrderStatusRequest {
    uint64 order_id = 1;
}

message GetTradeHistoryRequest {
    uint32 limit = 1;
}

message Trade {
    uint64 maker_order_id = 1;
    uint64 taker_order_id = 2;
    string price = 3;
    string quantity = 4;
    Side side = 5;
    google.protobuf.Timestamp timestamp = 6;
}

message TradeHistoryResponse {
    repeated Trade trades = 1;
}

message OrderBatchRequest {
    repeated OrderRequest orders = 1;
}

message OrderBatchResponse {
    repeated OrderResponse orders = 1;
}

enum Side {
    BID = 0;
    ASK = 1;
}

enum OrderType {
    LIMIT = 0;
    MARKET = 1;
}

enum OrderStatus {
    PENDING = 0;
    PARTIALLY_FILLED = 1;
    FILLED = 2;
    CANCELLED = 3;
}
